lib/gc/ExecutionEngine/CPURuntime/Microkernel/BrgemmOnednn.cpp

```diff
@@ -0,0 +1,156 @@
+//===-- BrgemmOnednn.cpp - BRGEMM Onednn Implementation ---------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <assert.h>
+#include <math.h>
+#include <mutex>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <vector>
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <unistd.h>
+#endif
+
+// manually include xbyak header here to avoid no-exception compile issue
+#define XBYAK_NO_EXCEPTION
+#include <cpu/x64/xbyak/xbyak.h> // NOLINT
+#undef XBYAK_NO_EXCEPTION
+
+#include <cpu/x64/amx_tile_configure.hpp>
+#include <cpu/x64/brgemm/brgemm.hpp>
+#include <cpu/x64/brgemm/brgemm_types.hpp>
+#include <cpu/x64/cpu_isa_traits.hpp>
+
+using namespace dnnl::impl::cpu::x64;
+
+namespace dnnl {
+namespace impl {
+namespace graph {
+namespace utils {
+// dummy definition for DNNL lite linkage
+__attribute__((weak)) void print_verbose_header() {}
+} // namespace utils
+} // namespace graph
+} // namespace impl
+} // namespace dnnl
+
+static constexpr int PALETTE_SIZE = 64;
+static std::mutex g_brgemm_mutex;
+static std::vector<brgemm_desc_t> g_brgemm_desc_list;
+static std::vector<brgemm_kernel_t *> g_brgemm_kernel_list;
+static std::vector<char *> g_brgemm_palette;
+
+// TODO(haixin): use syscall to determine page size?
+static constexpr size_t SCRATCH_SIZE = 2 * 4096;
+// TODO(haixin): need to use custom thread management for scratch in the future?
+static thread_local char scratch[SCRATCH_SIZE] = {0};
+
+extern "C" {
+
+int64_t dnnl_brgemm_dispatch(int64_t M, int64_t N, int64_t K, int64_t LDA,
+                             int64_t LDB, int64_t LDC, int64_t stride_a,
+                             int64_t stride_b, float beta, int64_t dtypeA,
+                             int64_t dtypeB) {
```

Comment:
Are we going to have a header with the entry point? So that the two (or whatever number of them we end up with) are guaranteed to not diverge.

Reply:
Entry point header added.